index.js

```diff
@@ -1,100 +1,253 @@
+  ctx.beginPath();
+  ctx.moveTo(0,titleHeight);
+  ctx.lineTo(canvasWidth, titleHeight);
+  ctx.stroke();
+  if(lootHeight && xpHeight){
+    ctx.beginPath();
+    ctx.moveTo(0,titleHeight + lootHeight);
+    ctx.lineTo(canvasWidth, titleHeight + lootHeight);
+    ctx.stroke();
+  }
+
+
+  //the loot
+  if(data?.loot?.length){
+    ctx.lineWidth = 2;
+    ctx.textAlign = "left";
+    ctx.font = "20px Runescape";
+    ctx.strokeText("Loot:", 15, 110);
+    ctx.fillText("Loot:", 15, 110);
+    let images = await getMultiple(data);
+    let itemsReady = await loadItems(images, ctx);
+  }
+
+  //the xp
+  if(data?.xp_earned){
+    ctx.lineWidth = 2;
+    ctx.font = "20px Runescape";
+    ctx.textAlign = "left";
+    ctx.strokeText("XP:", 15, 30+ titleHeight+lootHeight);
+    ctx.fillText("XP:", 15, 30 +titleHeight+lootHeight);
+    let skillsOpened = await openSkills(data.xp_earned);
+    let skillsLoaded = await loadSkills(skillsOpened, ctx, titleHeight + lootHeight);
+  }
+
+  //the prog report
   const response = {
     statusCode: 200,
-    body: JSON.stringify(canvas.toDataURL()),
+    body: JSON.stringify(canvas.toDataURL())
   };
-  
+
   return response;
 }
 
-async function loadImages(objs, data) {
-  let images = [];
-  await Promise.all(objs.map((obj,index) =>
-    loadImage(`data:image/png;base64,${obj}`).then((imageObj) => {
-      paintImages(imageObj, index, data);
+async function loadItems(objs, ctx) {
+  await Promise.all(objs.map((obj, index) =>
+    loadImage(`data:image/png;base64,${obj.img}`).then((imageObj) => {
+      paintIcons(ctx, imageObj, index, obj.count);
     })
   ));
-  return images;
 }
 
-function paintImages(image, index, data){
-    const offset = 40 * index;
-    ctx.drawImage(image, x + offset, 120);
-    ctx.font = "12px Runescape";
-    ctx.textAlign = "right";
-    const count = data.loot[index].count;
-    if (count >= 1000000) {
-    count = `${Math.floor(count / 1000000)}m`;
-    }
-    if (count > 1000) {
+async function openSkills(objs) {
+  let skills = [];
+  await Promise.all(objs.map((obj) => 
+    getSkill(`${__dirname}/icons/${obj.skill}.png`).then(response => {
+      skills.push({'img': `data:image/png;base64,${response.toString('base64')}`, 'xp': obj.xp});
+    })
+  ));
+  return skills;
+}
+
+const getSkill = async(file)=>{
+  const result = await fs.readFile(file)
+  return result;
+}
+
+async function loadSkills(objs, ctx, offset) {
+  await Promise.all(objs.map((obj, index) =>
+    loadImage(obj.img).then((imageObj) => {
+      paintSkills(ctx, imageObj, index, offset, `${obj.xp} xp`);
+    })
+  ));
+}
+
+function paintSkills(ctx, image, index, offset, xp){
+  const yOffset = offset + 40 + (50 * Math.floor(index / 5));
+  const xOffset = 15 + (60 * (index % 5));
+  ctx.drawImage(image, xOffset, yOffset, 25, 25);
+  ctx.font = "16px Runescape";
+  ctx.textAlign = "center";
+  ctx.fillStyle = "#ffff00";
+  const xpInt = parseInt(xp.replace(/,/g, ''));
+  let xpGained = '';
```

Comment:
```
  let xpGained = 0;
  let xpUnit = '';
  if (xpInt >= 1000000) {
    xpGained = Math.floor(xpInt * 10 / 1000000) / 10;
    xpUnit ='m';
  }
  else if (xpInt > 1000) {
    xpGained = Math.trunc((xpInt * 10) / 1000 / 10);
    xpUnit ='k';
  }
  
  `${xpGained}${xpUnit} ${xpUnit}`
```
or not :D

Reply:
I forgot to say why: It seems like encapsulating this math into a template is silly. Another alterative is to use moooooar templates...

Reply:
```
 `${(
     xpInt >= 1000000 ? Math.floor(xpInt * 10 / 1000000) / 10 : (
       xpInt > 1000 ? Math.trunc((xpInt * 10) / 1000 / 10); : xpInt
     )
   )}${
   (
     xpInt >= 1000000 ? 'm' : (
       xpInt > 1000 ? 'k' : ''
     )
   )} ${xpUnit}`
```
lol... dont do that. Also in the first example 

```
  let xpGained = 0;
```
should be
```
  let xpGained = xpInt;
```